### DISCUSSION TITLE
Multiple custom event handlers

### DISCUSSION THREAD

**Discussion by ghost** [rabbibotton/clog#225] [General]
Hi. Is the intention to allow multiple event handlers to be triggered for the same custom event. I have noticed that while one can set *n* handlers for a custom event via `set-on-event` and `set-on-event-with-data` only one of the handlers is chosen when triggering a custom event (the last one that was set). This is despite *n* handlers being triggered, in effect triggering *n* of the same handler.

I think it would be nice to allow this as it would make it possible to decouple components, giving the ability to send outside events to a proxy object while allowing sub-components of that object to specify their own behavior.

For instance lets say there is a numpad component with nine keys 0-9, I would like for the numpad to highlight one of its keys if the user has pressed a number key on their keyboard. A nice way to do this would have the numpad receiving a custom event 'user-input' w/ a string that represents the key that the user just pressed on their keyboard. This way, during creation of the numpad, the sub-components, the keys/key caps, can register event handlers for 'user-input' on behalf of the numpad component, changing their appearance as required, making sure to check if the event was intended for themselves first.

It would be completely opt-in (one would have to explicitly set multiple handlers for the same event) but it would mean that for every 'user-event' triggered a total of nine handlers are triggered, with all but one immediately exiting.

**Comment by rabbibotton** [comment]
I would not use the html page as the means to pass messages. While I do so in rare cases, they have almost always been flags that there is a design flaw in the application, as there is a "toll" to pay to pass messages between the page and the application. Once an event has been received on the Lisp side, better to choose a Lisp way of dealing with it.

Let's consider this number pad:

```
(defun my-page (body)
  (clog-web:clog-web-initialize body)
  (add-class body "w3-dark-grey")  
  (let* ((calc         (create-div body :class "w3-monospace w3-xxlarge w3-center"))
         (display      (create-div calc :class "w3-center" :content "0"))  
         (button-class "w3-border w3-margin w3-padding w3-ripple"))
    (labels ((on-click (obj)
               (setf (text display) (text obj)))
             (on-key (obj data)
               (let ((key-code (getf data :key-code))
                     (key      (getf data :key)))
                 (when (<= 48 key-code 57)
                   (setf (text display) key)))))
      (set-on-key-press body #'on-key)
      (dotimes (row 3)
        (dotimes (col 3)
          (let* ((loc (+ (+ (* row 3) col) 1))
                 (key-btn (create-span calc :content loc :class button-class)))
            (set-on-click key-btn #'on-click)))
        (create-br calc))
      (create-span calc :content "&nbsp;" :class button-class)
      (set-on-click (create-span calc :content 0 :class button-class) #'on-click)
      (create-span calc :content "&nbsp;" :class button-class))))
```

There is no custom messages, and only 1 handler defined. the event's lambda function receives the key-btn that was pressed as its first parameter. (In truth there are two since we are capturing keyboard too)

I would not want (and certainly not depend it) to let's say one key sending two on-click message to two different handlers. If I pass to any handler a nil handler all handlers are turned off and this still means multiple "tolls" to pay for each handler.

Instead if I wanted to dispatch and on-click to multiple handers, I would just push handlers on to a list and mapcar to fire each one on the lisp side in the one on-click-handler. This method has no "toll" and is optionally installed with no real effort.

Think in Lisp not in HTML/JS, I figure since Lisp older than I, it is the wiser choice :)

If I am misunderstanding you question, please let me know.

**Comment by rabbibotton** [comment]
So here is that same sample but with the background of keys flashing red (clicked or num keys) using a hash.

```
(defun my-page (body)
  (clog-web:clog-web-initialize body)
  (add-class body "w3-dark-grey")  
  (let* ((calc         (create-div body :class "w3-monospace w3-xxlarge w3-center"))
         (display      (create-div calc :class "w3-center" :content "0"))  
         (button-class "w3-border w3-margin w3-padding w3-ripple")
         (key-hash     (make-hash-table*))) ;; * to use thread safe version
    (labels ((on-click (obj)
               (setf (text display) (text obj))
               (setf (background-color obj) :red)
               (sleep .1)
               (setf (background-color obj) ""))
             (on-key (obj data)
               (let ((key-code (getf data :key-code))
                     (key      (getf data :key)))
                 (when (<= 48 key-code 57)                   
                   (setf (text display) key)
                   (let ((obj (gethash (- key-code 48) key-hash)))
                     (setf (background-color obj) :red)
                     (sleep .1)
                     (setf (background-color obj) ""))))))
      (set-on-key-press body #'on-key)
      (dotimes (row 3)
        (dotimes (col 3)
          (let* ((loc (+ (+ (* row 3) col) 1))
                 (key (create-span calc :content loc :class button-class)))
            (setf (gethash loc key-hash) key)
            (set-on-click key #'on-click)))
        (create-br calc))
      (create-span calc :content "&nbsp;" :class button-class)
      (let ((key (create-span calc :content 0 :class button-class)))
        (setf (gethash 0 key-hash) key)
        (set-on-click key #'on-click))
      (create-span calc :content "&nbsp;" :class button-class))))
```

**Comment by ghost** [comment]
My main concern is what to do when parts become decoupled from each other. I have been away from the keyboard for a day or two, but on thinking about it more, I realise I am misappropriating the custom eventing mechanism in Clog to achieve decoupling (It also doesn't help that I have just picked up programming again after some time away). Nevertheless, the question is how to design a protocol for parts to talk to each other.

As soon as the creation of a component occurs in a context apart from the 'main' context visibility is lost, and there is no way to reference subcomponents (elements) with out (1) drawing references to those subcomponents back to the 'main' context (via return values, probably), or (2) using a mutually agreed-upon convention to somehow signal one another either uni-directionally or bi-directionally.

There is a downside to (1), and that is one most likely wants references to subcomponents so that behavior can be specified on certain events. This means, though, that the component isn't truly decoupled since the component creation is separate from the specification of behavior. So it leaves (2) as the desired strategy (for me, at least).

I have been attempting to use the custom eventing mechanism as an example of the latter strategy. It seemed desirable since clog can hold references to subcomponents inside the closure that is the event handler, the handlers are bound to an event ('numkey-1'... 'numkey-9' for example), and then triggered from any context as long as it contains a reference to the proxy object (numpad). Though, it's not the best way, as there is a round-trip that is paid when triggering custom events. Since learning Clog I have been trying to see how I can use all it's parts to design apps, but this is not the appropriate use as you have said.

Today I worked on a different method to allow the decoupling of components. If you have the time can you tell me if I am on the right track or not?

```
(defclass numpad-component (clog-element)
  ((numkey-handlers
    :initform (make-hash-table* :test #'equal)
    :accessor numkey-handlers)))

(defmethod register-numkey-handler ((numpad numpad-component) key handler)
  (setf (gethash key (numkey-handlers numpad)) handler))

(defmethod send ((numpad numpad-component) key)
  (let ((key-handler (gethash key (numkey-handlers numpad))))
    (unless (null key-handler)
      (bordeaux-threads:make-thread key-handler))))

(defmethod create-numpad
    ((obj clog-obj)
     &key
       (content "")
       (style nil)
       (hidden nil)
       (numpad-class "w-100 pa2 flex flex-wrap flex-row-reverse justify-center f2 tc code")
       (outer-keycap-class "pa2")
       (inner-keycap-class "grow pointer bg-near-black white w3 h3 br3 flex items-center justify-center")
       (html-id nil)
       (auto-place t)
       (continuation nil))
  (let* ((numpad (create-child obj (format nil "<div~@[~A~]~@[~A~]>~A</div>"
                                           (when numpad-class
                                             (format nil " class='~A'"
                                                     (escape-string numpad-class :html t)))
                                           (when (or hidden style)
                                             (format nil " style='~@[~a~]~@[~a~]'"
                                                     (when hidden "visibility:hidden;")
                                                     style))
                                           content)
                               :clog-type  'numpad-component
                               :html-id    html-id
                               :auto-place auto-place)))
    (loop for i from 9 downto 0
          do (let* ((key-number i)
                    (key-symbol (format nil "~A" key-number))
                    (outer-keycap (create-div numpad :class outer-keycap-class))
                    (inner-keycap (create-div outer-keycap :class inner-keycap-class)))
               (create-span inner-keycap :content key-symbol)

               ;; Register the keycap handler w/ numpad
               ;; Rate limiting and thread-safety is achieved by
               ;; the use of a lock
               (register-numkey-handler
                numpad key-symbol
                (let ((rate-limiter (bordeaux-threads:make-lock)))
                  (lambda ()
                    (when (bordeaux-threads:acquire-lock rate-limiter nil)
                      (add-class inner-keycap "bg-dark-gray")
                      (sleep 0.250)
                      (remove-class inner-keycap "bg-dark-gray")
                      (bordeaux-threads:release-lock rate-limiter)))))

               ;; Register on-click event for keycap
               (set-on-click
                inner-keycap
                (lambda (obj)
                  (declare (ignore obj))
                  (send numpad key-symbol)
                  (unless (null continuation)
                    (funcall continuation key-symbol))))))
    numpad))

(defun untitled (body)
  (setf (title (html-document body)) "Untitled")
  (load-css (html-document body) "https://unpkg.com/tachyons@4.12.0/css/tachyons.min.css")
  (add-class (body-element (html-document body)) "bg-near-white near-black")
  (let* ((main-container (create-div body :class "mw5 mt5 center sans-serif"))
         (outer-display (create-div main-container :class "w-100 h4 flex flex-row justify-center items-center f1"))
         (chunk-display-line (create-div outer-display :class "bb bw1"))
         (chunk-display (create-div chunk-display-line :class "code" :content "&nbsp;")))
    (labels ((write-chunk-to-display (string)
               (setf (text chunk-display) string)))
      (let ((numpad (create-numpad main-container :continuation #'write-chunk-to-display)))
        (labels ((key-press-handler (obj data)
                   (declare (ignore obj))
                   (let ((key-code (getf data :key-code))
                         (key (getf data :key)))
                     (when (<= 48 key-code 57)
                       (send numpad key)
                       (write-chunk-to-display key)))))
          (set-on-key-press body #'key-press-handler))))))

(defun launch (&key (host "0.0.0.0"))
  (clog:initialize 'untitled :host host)
  (clog:open-browser))
```

Thanks for your help, and your patience :)

**Reply by rabbibotton** [reply]
Looks good :) Check out clog:with-sync-event if you use that on your event handlers for the mouse and keyboard it will give you a single queue of events and can remove your need for locks.

**Reply by ghost** [reply]
Okay cheers.

Also I found it's much cleaner to do `(change-class (create-div obj ... ) 'component-class)`. Keeps it DRY too, so modifications to Clog element creation wont require modifying components at a later date.

**Reply by rabbibotton** [reply]
Take a look at any panel code generated by the builder.

**Reply by ghost** [reply]
Sure, I'll take a look, so far I've yet to try using the builder.
